Examples/Io/Root/src/RootTrajectoryWriter.cpp

```diff
@@ -271,36 +299,40 @@ ActsExamples::ProcessCode ActsExamples::RootTrajectoryWriter::writeT(
   m_eventNr = ctx.eventNumber;
 
   // Loop over the trajectories
-  int iTraj = 0;
-  for (const auto& traj : trajectories) {
-    m_trajNr = iTraj;
+  for (size_t itraj = 0; itraj < trajectories.size(); ++itraj) {
+    const auto& traj = trajectories[itraj];
 
-    // The trajectory entry indices and the multiTrajectory
-    const auto& [trackTips, mj] = traj.trajectory();
-    if (trackTips.empty()) {
-      ACTS_WARNING("Empty multiTrajectory.");
+    if (traj.empty()) {
+      ACTS_WARNING("Empty trajectories object " << itraj);
       continue;
     }
+
+    m_trajNr = itraj;
+
+    // The trajectory entry indices and the multiTrajectory
+    const auto& mj = traj.multiTrajectory();
+    const auto& trackTips = traj.tips();
     // Check the size of the trajectory entry indices. For track fitting, there
     // should be at most one trajectory
     if (trackTips.size() > 1) {
       ACTS_ERROR("Track fitting should not result in multiple trajectories.");
       return ProcessCode::ABORT;
     }
-    // Get the entry index for the single trajectory
-    auto& trackTip = trackTips.front();
 
+    // Get the entry index for the single trajectory
+    auto trackTip = trackTips.front();
     // Collect the trajectory summary info
     auto trajState =
         Acts::MultiTrajectoryHelpers::trajectoryState(mj, trackTip);
     m_nMeasurements = trajState.nMeasurements;
     m_nStates = trajState.nStates;
 
     // Get the majority truth particle to this track
-    const auto particleHitCount = traj.identifyMajorityParticle(trackTip);
-    if (not particleHitCount.empty()) {
+    identifyContributingParticles(hitParticlesMap, traj, trackTip,
```

Comment:
I doubt whether we really need the `hitParticlesMap` while we already have the `hitSimHitsMap` for identifying the majority particle. We might overload the function `identifyContributingParticles` to use the `hitSimHitsMap` as well. In this way, we could simplify the input collections a bit.

Reply:
In principle that would be possible. However, then the track classifiers would need both the `hitSimHitsMap` and the `simHits` containers as input and you need to do a double lookup, i.e. first find all `SimHit`s for a measured hit and then access the particle for a `SimHit`. Since we are iterating over all `SimHits` anyway during digitization, we can create the additional map basically for free. While the output collections e.g. for the digitizers become simpler, the classifications becomes more complicated (and I suspect slower).

On a completely unrelated note: do you think the naming is clear enough? I fear that there can be confusion between hit and simhit: the first describes the measured one while the later describes the simulation output. I am considering to use `Measurement` consistently to describe hits after digitization and e.g. use `measurementParticlesMap` and `measurementSimHitsMap` instead.

Reply:
Ah, right. By the `hitSimHitsMap` alone, it's not possible to access the `simHit` information. So it's better to use `hitParticlesMap`.

I think naming with explicit `measurement` might sound more clear since we all know `hit` is a confusing word.

Reply:
I renamed the output collections to `measurement...` to avoid confusion.